File: src/agent.py
# """
# AI Interview Agent - Pure Gemini Multimodal implementation
# Main agent implementation using LiveKit and Gemini Multimodal Live API
# """

# import os
# import json
# import logging
# import asyncio
# from datetime import datetime
# from pathlib import Path
# from dotenv import load_dotenv

# # Load environment variables
# project_root = Path(__file__).parent.parent
# env_path = project_root / '.env'
# load_dotenv(env_path)

# from livekit.agents import (
#     Agent,
#     AgentSession,
#     JobContext,
#     WorkerOptions,
#     cli,
#     tokenize,
# )
# from livekit.plugins import google
# from livekit import rtc

# from prompts import get_interview_prompt
# from local_recording_manager import LocalRecordingManager
# from transcription_handler import TranscriptionHandler

# # Configure logging
# logging.basicConfig(
#     level=logging.INFO,
#     format='%(asctime)s - %(name)s - %(levelname)s - %(message)s'
# )
# logger = logging.getLogger(__name__)

# class InterviewAgent:
#     """Main interview agent class using Gemini Multimodal Live"""
    
#     def __init__(
#         self,
#         candidate_id: str,
#         job_role: str = "software_engineer",
#         interview_duration: int = 30
#     ):
#         self.candidate_id = candidate_id
#         self.job_role = job_role
#         self.interview_duration = interview_duration
#         self.start_time = None
#         self.recorder = None
#         self.transcription = None

#     async def start_interview(self, ctx: JobContext):
#         """Start the interview session"""
#         logger.info(f"--- STARTING MULTIMODAL INTERVIEW (Candidate: {self.candidate_id}) ---")
        
#         try:
#             # 1. Initialize components
#             api_key = os.getenv("GOOGLE_API_KEY")
#             interview_instructions = get_interview_prompt(self.job_role)
            
#             # Using gemini-2.5-flash-native-audio-latest which is verified to work!
#             logger.info("Initializing RealtimeModel (Gemini 2.5 Flash Native Audio)...")
#             model = google.realtime.RealtimeModel(
#                 model="gemini-2.5-flash-native-audio-latest",
#                 api_version="v1beta",
#                 api_key=api_key,
#                 instructions=interview_instructions,
#                 voice="Puck",
#             )

#             # Create the modern Agent holding the model
#             logger.info("Initializing Agent and AgentSession...")
#             self.agent = Agent(llm=model, instructions=interview_instructions)
#             self.session = AgentSession()
            
#             # 2. Setup Recording & Transcription
#             self.recorder = LocalRecordingManager()
#             self.transcription = TranscriptionHandler(
#                 interview_id=f"{self.candidate_id}_{datetime.now().strftime('%Y%m%d_%H%M%S')}"
#             )

#             # 3. Add Event Handlers for Transcription (Robustness)
#             @self.session.on("user_input_transcribed")
#             def _on_user_transcript(ev):
#                 if ev.transcript:
#                     logger.info(f"--- USER: {ev.transcript} ---")

#             @self.session.on("conversation_item_added")
#             def _on_item_added(ev):
#                 # ev.item is a ChatMessage
#                 if ev.item.role == "assistant" and ev.item.text_content:
#                     logger.info(f"--- BOT: {ev.item.text_content} ---")
#                     if self.transcription:
#                         asyncio.create_task(self.transcription.on_transcript(
#                             type('Event', (), {'participant': type('P', (), {'identity': 'AI_Interviewer'})(), 'text': ev.item.text_content, 'is_final': True})
#                         ))

#             # 4. Connect and Start
#             await ctx.connect()
#             logger.info(f"Connected to room: {ctx.room.name}")

#             # Start local recording (subscribes to tracks)
#             await self.recorder.start_recording(ctx.room, self.candidate_id)

#             # Start the session (Modern API requires the agent)
#             # Setting record=False to avoid 401 errors in LiveKit Cloud
#             await self.session.start(self.agent, room=ctx.room, record=False)
            
#             self.start_time = datetime.now()
#             logger.info("AgentSession started.")

#             # Greet the candidate using the RealtimeModel's native generation
#             # Since .say() requires a TTS, we push a message to history and generate
#             self.session.history.add_message(
#                 role="assistant",
#                 content="Hi! I'm your AI interviewer. How are you doing today?"
#             )
#             self.session.generate_reply()

#             # 5. Monitor
#             await self._monitor_interview_duration(ctx)
            
#         except Exception as e:
#             logger.error(f"Error in multimodal interview: {e}")
#             raise
#         finally:
#             # CLEANUP: Ensure recordings and transcripts are saved
#             logger.info("Multimodal session cleanup starting...")
#             if self.recorder:
#                 try:
#                     info = self.recorder.stop_recording()
#                     logger.info(f"Local recording stopped: {info}")
#                 except: pass
            
#             if self.transcription:
#                 try:
#                     transcript_text = self.transcription.get_full_transcript()
#                     os.makedirs("transcripts", exist_ok=True)
#                     path = f"transcripts/{self.candidate_id}_transcript.txt"
#                     with open(path, "w") as f:
#                         f.write(transcript_text)
#                     logger.info(f"Final transcript saved to {path}")
#                 except: pass
            
#             logger.info("Cleanup complete.")

#     async def _monitor_interview_duration(self, ctx: JobContext):
#         max_duration_seconds = self.interview_duration * 60
#         while True:
#             await asyncio.sleep(60)
#             if self.start_time:
#                 elapsed = (datetime.now() - self.start_time).total_seconds()
#                 if elapsed >= max_duration_seconds:
#                     logger.info("Interview limit reached.")
#                     await ctx.room.disconnect()
#                     break

# async def entrypoint(ctx: JobContext):
#     logger.info(f"JOB RECEIVED: {ctx.job.id}")
    
#     # Get metadata
#     room_metadata = json.loads(ctx.room.metadata) if ctx.room.metadata else {}
#     candidate_id = room_metadata.get("candidate_id", "unknown")
#     job_role = room_metadata.get("job_role", "software_engineer")
    
#     agent = InterviewAgent(candidate_id=candidate_id, job_role=job_role)
#     await agent.start_interview(ctx)

# if __name__ == "__main__":
#     cli.run_app(
#         WorkerOptions(
#             entrypoint_fnc=entrypoint,
#         )
#     )

"""
AI Interview Agent - Pure Gemini Multimodal implementation
Main agent implementation using LiveKit and Gemini Multimodal Live API
"""
import os
import json
import logging
import asyncio
from datetime import datetime
from pathlib import Path
from dotenv import load_dotenv

# Load environment variables
project_root = Path(__file__).parent.parent
env_path = project_root / '.env'
load_dotenv(env_path)

from livekit.agents import (
    Agent,
    AgentSession,
    JobContext,
    WorkerOptions,
    cli,
    tokenize,
)
from livekit.plugins import google
from livekit import rtc

from prompts import get_interview_prompt
from cloud_recording_manager import CloudRecordingManager
from transcription_handler import TranscriptionHandler

# Configure logging
logging.basicConfig(
    level=logging.INFO,
    format='%(asctime)s - %(name)s - %(levelname)s - %(message)s'
)
logger = logging.getLogger(__name__)

class InterviewAgent:
    """Main interview agent class using Gemini Multimodal Live"""
    
    def __init__(
        self,
        candidate_id: str,
        room_id: str,
        job_role: str = "software_engineer",
        interview_duration: int = 30
    ):
        self.candidate_id = candidate_id
        self.room_id = room_id
        self.job_role = job_role
        self.interview_duration = interview_duration
        self.start_time = None
        self.recorder = None
        self.transcription = None
        self.is_cleaning_up = False

    async def handle_cleanup(self, ctx: JobContext):
        """Handle graceful shutdown, stop recording, and notify frontend"""
        if self.is_cleaning_up:
            return
        self.is_cleaning_up = True
        
        logger.info("=" * 60)
        logger.info("STARTING CLEANUP AND SAVING RECORDINGS...")
        logger.info("=" * 60)
        
        # Stop cloud recording
        download_url = None
        if self.recorder:
            try:
                logger.info("Stopping cloud recording...")
                egress = await self.recorder.stop_recording()
                if egress:
                    logger.info(f"✓ Cloud recording process final status: {egress.status}")
                    if egress.file_results and len(egress.file_results) > 0:
                        download_url = egress.file_results[0].download_url
                        if download_url:
                            logger.info(f"Download available at: {download_url}")
                    else:
                        logger.info("Recording is processing. Check LiveKit Console.")
            except Exception as e:
                logger.error(f"Error stopping cloud recording: {e}", exc_info=True)
            
            # Additional small delay to ensure LiveKit processes the stop request
            logger.info("Waiting for recording finalization...")
            await asyncio.sleep(2)

        # Send Data Packet to frontend (to trigger popup/notification)
        try:
            payload = json.dumps({
                "type": "INTERVIEW_ENDED",
                "status": "completed",
                "room_id": self.room_id,
                "recording_url": download_url,
                "message": "Interview completed. Recording is processing."
            }).encode('utf-8')

            if ctx.room.local_participant:
                await ctx.room.local_participant.publish_data(
                    payload,
                    topic="interview_status"
                )
                logger.info("✓ Sent INTERVIEW_ENDED data packet to frontend")
                # Wait a bit for packet to be delivered
                await asyncio.sleep(2)
        except Exception as e:
            logger.error(f"Failed to send data packet: {e}")
        
        # Save transcription
        if self.transcription:
            logger.info("Saving transcription and uploading to S3...")
            # This triggers the S3 upload logic now built into save_formatted_transcript
            egress_id = self.recorder.egress_id if self.recorder else None
            self.transcription.save_formatted_transcript(egress_id=egress_id)
        
        # Disconnect
        logger.info("Disconnecting room...")
        await ctx.room.disconnect()
        
    async def start_interview(self, ctx: JobContext):
        """Start the interview session"""
        logger.info(f"--- STARTING MULTIMODAL INTERVIEW (Candidate: {self.candidate_id}) ---")
        
        try:
            # 1. Initialize components
            api_key = os.getenv("GOOGLE_API_KEY")
            interview_instructions = get_interview_prompt(self.job_role)
            
            # Using gemini-2.5-flash-native-audio-latest which is verified to work!
            logger.info("Initializing RealtimeModel (Gemini 2.5 Flash Native Audio)...")
            model = google.realtime.RealtimeModel(
                model="gemini-2.5-flash-native-audio-latest",
                api_version="v1beta",
                api_key=api_key,
                instructions=interview_instructions,
                voice="Puck",
            )
            
            # Create the modern Agent holding the model
            logger.info("Initializing Agent and AgentSession...")
            self.agent = Agent(llm=model, instructions=interview_instructions)
            self.session = AgentSession()
            
            # 2. Setup Recording & Transcription
            # Use Room ID for S3 folder structure
            self.recorder = CloudRecordingManager()
            self.transcription = TranscriptionHandler(
                interview_id=self.room_id 
            )
            
            # 3. Add Event Handlers for Transcription (Robustness)
            @self.session.on("user_input_transcribed")
            def _on_user_transcript(ev):
                if ev.transcript:
                    logger.info(f"--- USER: {ev.transcript} ---")
                    if self.transcription:
                        asyncio.create_task(self.transcription.on_transcript(
                            type('Event', (), {
                                'participant': type('P', (), {'identity': self.candidate_id})(),
                                'text': ev.transcript,
                                'is_final': True
                            })
                        ))
            
            @self.session.on("conversation_item_added")
            def _on_item_added(ev):
                # ev.item is a ChatMessage
                if ev.item.role == "assistant" and ev.item.text_content:
                    logger.info(f"--- BOT: {ev.item.text_content} ---")
                    if self.transcription:
                        asyncio.create_task(self.transcription.on_transcript(
                            type('Event', (), {
                                'participant': type('P', (), {'identity': 'AI_Interviewer'})(),
                                'text': ev.item.text_content,
                                'is_final': True
                            })
                        ))
            
            # 4. Connect and Start
            # Auto-subscribe is enabled by default in newer SDK versions
            logger.info("Connecting to room...")
            await ctx.connect()
            logger.info(f"Connected to room: {ctx.room.name}")
            
            # Subscribe to data packets (for user signals like "End Call")
            @ctx.room.on("data_received")
            def on_data_received(payload: bytes, participant: rtc.RemoteParticipant, kind: rtc.DataPacketKind, topic: str = ""):
                if topic == "user_actions":
                    try:
                        data = json.loads(payload.decode('utf-8'))
                        if data.get("type") == "USER_ENDED_CALL":
                            logger.info("Received USER_ENDED_CALL signal")
                            asyncio.create_task(self.handle_cleanup(ctx))
                    except Exception as e:
                        logger.error(f"Failed to process data packet: {e}")

            # Handle user disconnecting (closing tab)
            @ctx.room.on("participant_disconnected")
            def on_participant_disconnected(participant: rtc.RemoteParticipant):
                logger.info(f"Participant disconnected: {participant.identity}")
                # If the candidate leaves, we should wrap up
                if participant.identity == self.candidate_id or "candidate" in participant.identity.lower():
                    logger.info("Candidate left the interview. Cleaning up...")
                    asyncio.create_task(self.handle_cleanup(ctx))

            # Manually subscribe to all remote tracks to ensure recording works
            for participant in ctx.room.remote_participants.values():
                for publication in participant.track_publications.values():
                    if not publication.subscribed and publication.track:
                        publication.set_subscribed(True)
                        logger.info(f"Subscribed to {publication.kind} track from {participant.identity}")
            
            # Start local recording (will now receive tracks properly)
            # Start Cloud Recording (Egress)
            try:
                logger.info("Starting cloud recording...")
                await self.recorder.start_recording(ctx.room.name, self.room_id)
            except Exception as e:
                logger.error(f"FAILED TO START CLOUD RECORDING: {e}")
                
                # Notify in chat so user knows recording isn't happening
                if self.session: 
                    self.session.history.add_message(
                        role="assistant",
                        content="Warning: Session recording could not be started. Please check server configuration."
                    )
            
            # Subscribe to tracks as they're published
            @ctx.room.on("track_published")
            def on_track_published(publication: rtc.RemoteTrackPublication, participant: rtc.RemoteParticipant):
                logger.info(f"New track published: {publication.kind} from {participant.identity}")
                publication.set_subscribed(True)
            
            # Start the session (Modern API requires the agent)
            # Setting record=False to avoid 401 errors in LiveKit Cloud
            logger.info("Starting agent session...")
            await self.session.start(self.agent, room=ctx.room, record=False)
            
            self.start_time = datetime.now()
            logger.info("AgentSession started successfully.")
            
            # Greet the candidate using the RealtimeModel's native generation
            logger.info("Sending initial greeting...")
            self.session.history.add_message(
                role="assistant",
                content="Hi! I'm your AI interviewer. How are you doing today?"
            )
            self.session.generate_reply()
            
            # 5. Monitor interview duration
            await self._monitor_interview_duration(ctx)
            
        except Exception as e:
            logger.error(f"Error in multimodal interview: {e}", exc_info=True)
            raise
        finally:
            # CLEANUP: Ensure recordings and transcripts are saved
            await self.handle_cleanup(ctx)
    
    async def _monitor_interview_duration(self, ctx: JobContext):
        """Monitor interview duration and disconnect when time limit reached"""
        max_duration_seconds = self.interview_duration * 60
        
        while True:
            await asyncio.sleep(30)  # Check every 30 seconds
            
            if self.start_time:
                elapsed = (datetime.now() - self.start_time).total_seconds()
                remaining = max_duration_seconds - elapsed
                
                # Log remaining time every 5 minutes
                if int(elapsed) % 300 == 0:
                    logger.info(f"Interview progress: {int(elapsed/60)} minutes elapsed, {int(remaining/60)} minutes remaining")
                
                # End interview when time limit reached
                if elapsed >= max_duration_seconds:
                    logger.info("=" * 60)
                    logger.info("INTERVIEW TIME LIMIT REACHED")
                    logger.info("=" * 60)
                    
                    # Send final message to candidate
                    try:
                        self.session.history.add_message(
                            role="assistant",
                            content="Thank you for your time! The interview has concluded. Have a great day!"
                        )
                        self.session.generate_reply()
                        await asyncio.sleep(3)  # Give time for message to be sent
                    except:
                        pass
                    
                    # Disconnect
                    await ctx.room.disconnect()
                    break


async def entrypoint(ctx: JobContext):
    """Main entrypoint for the agent"""
    logger.info("=" * 60)
    logger.info(f"NEW JOB RECEIVED: {ctx.job.id}")
    logger.info(f"Room: {ctx.room.name}")
    logger.info("=" * 60)
    
    # Get metadata from room
    room_metadata = json.loads(ctx.room.metadata) if ctx.room.metadata else {}
    candidate_id = room_metadata.get("candidate_id", "unknown")
    job_role = room_metadata.get("job_role", "software_engineer")
    print(room_metadata)
    
    # FALLBACK: Ensure we never use "unknown" for folders
    if candidate_id == "unknown":
        import uuid
        # Generate a unique session ID
        candidate_id = f"session-{datetime.now().strftime('%Y%m%d_%H%M%S')}-{str(uuid.uuid4())[:6]}"
        logger.info(f"Metadata missing. Generated fallback ID: {candidate_id}")
    
    logger.info(f"Candidate ID: {candidate_id}")
    logger.info(f"Job Role: {job_role}")
    
    room_id = ctx.room.sid
    
    # Create and start interview agent
    agent = InterviewAgent(
        candidate_id=candidate_id,
        room_id=room_id,
        job_role=job_role,
        interview_duration=30  # 30 minutes default
    )
    
    await agent.start_interview(ctx)


if __name__ == "__main__":
    cli.run_app(
        WorkerOptions(
            entrypoint_fnc=entrypoint,
        )
    )

    # //upto candidate audio recording 

File: backend/agent.js
/**
 * AI Interview Agent - Node.js Implementation
 * Uses LiveKit Agents Framework (1.0+) and Gemini Multimodal Live API
 */

import { cli, defineAgent, voice, WorkerOptions } from '@livekit/agents';
import { RoomEvent } from '@livekit/rtc-node';
import * as google from '@livekit/agents-plugin-google';
import path from 'path';
import { fileURLToPath } from 'url';
import dotenv from 'dotenv';
import { getInterviewPrompt } from './prompts.js';
import { CloudRecordingManager } from './cloud-recording-manager.js';
import { TranscriptionHandler } from './transcription-handler.js';

const __filename = fileURLToPath(import.meta.url);
const __dirname = path.dirname(__filename);

// Load env vars explicitly so the agent can find them no matter where it is launched from.
const rootEnv = dotenv.config({ path: path.join(__dirname, '../.env') });
const backendEnv = dotenv.config({ path: path.join(__dirname, '.env') });

const pickFirstNonEmpty = (...values) => {
    for (const value of values) {
        if (typeof value !== 'string') continue;

        const trimmed = value.trim();
        if (trimmed) return trimmed;
    }

    return undefined;
};

const resolveGoogleApiKey = () => {
    const candidates = [
        { source: 'backend/.env GOOGLE_API_KEY', value: backendEnv.parsed?.GOOGLE_API_KEY },
        { source: 'backend/.env GOOGLE_GENAI_API_KEY', value: backendEnv.parsed?.GOOGLE_GENAI_API_KEY },
        { source: 'backend/.env GEMINI_API_KEY', value: backendEnv.parsed?.GEMINI_API_KEY },
        { source: 'root .env GOOGLE_API_KEY', value: rootEnv.parsed?.GOOGLE_API_KEY },
        { source: 'root .env GOOGLE_GENAI_API_KEY', value: rootEnv.parsed?.GOOGLE_GENAI_API_KEY },
        { source: 'root .env GEMINI_API_KEY', value: rootEnv.parsed?.GEMINI_API_KEY },
        { source: 'process.env GEMINI_API_KEY', value: process.env.GEMINI_API_KEY },
        { source: 'process.env GOOGLE_API_KEY', value: process.env.GOOGLE_API_KEY },
        { source: 'process.env GOOGLE_GENAI_API_KEY', value: process.env.GOOGLE_GENAI_API_KEY },
    ];

    const picked = candidates.find(({ value }) => typeof value === 'string' && value.trim());
    const apiKey = picked?.value?.trim();

    if (!apiKey) {
        throw new Error(
            'Google API key not found. Set GOOGLE_API_KEY, GOOGLE_GENAI_API_KEY, or GEMINI_API_KEY in your .env file.',
        );
    }

    console.log(`Resolved Gemini API key from ${picked.source}`);

    // Keep the resolved key in process.env so downstream SDKs can read it too.
    process.env.GOOGLE_API_KEY = apiKey;
    process.env.GOOGLE_GENAI_API_KEY = apiKey;
    process.env.GEMINI_API_KEY = apiKey;
    return apiKey;
};

export default defineAgent({
    entry: async (ctx) => {
        console.log("=".repeat(60));
        console.log("NEW JOB RECEIVED");
        console.log("=".repeat(60));

        // STEP 1: Connect to the room.
        // ctx.connect() MUST be called before any room operations.
        // Without this, ctx.waitForParticipant() throws "room is not connected".
        await ctx.connect();
        console.log("✅ Connected to room:", ctx.room.name);

        // STEP 2: Wait for the candidate to join
        const participant = await ctx.waitForParticipant();
        console.log(`✅ Participant joined: ${participant.identity}`);

        // STEP 3: Extract candidate info from participant metadata
        let candidateId = participant.identity || 'unknown';
        let jobRole = 'software_engineer';

        if (participant.metadata) {
            try {
                const meta = JSON.parse(participant.metadata);
                candidateId = meta.candidate_id || candidateId;
                jobRole = meta.job_role || jobRole;
            } catch (e) {
                console.warn("Could not parse participant metadata, using defaults");
            }
        }

        console.log(`Candidate: ${candidateId} | Role: ${jobRole} | Room: ${ctx.room.name}`);

        // STEP 4: Setup session identifiers
        const timestamp = new Date().toISOString().replace(/[:.]/g, '-');
        const uniqueSessionId = `session_${timestamp}`;
        const interviewDuration = 30; // minutes

        const recorder = new CloudRecordingManager();
        const transcription = new TranscriptionHandler(`${candidateId}_${uniqueSessionId}`);

        // STEP 5: Build the Gemini Realtime model.
        // IMPORTANT: Must use a Gemini Live API model (bidiGenerateContent).
        // "gemini-2.0-flash-exp" was deprecated; use the preview below.
        // SDK default (non-Vertex AI): "gemini-2.5-flash-native-audio-preview-12-2025"
        const instructions = getInterviewPrompt(jobRole);
        const googleApiKey = resolveGoogleApiKey();

        const model = new google.beta.realtime.RealtimeModel({
            model: "gemini-2.5-flash-native-audio-preview-12-2025",
            apiKey: googleApiKey,
            instructions: instructions,
            voice: "Puck",
        });

        // STEP 6: Create the Agent (llm goes on Agent, agent_activity picks it up)
        const agent = new voice.Agent({
            instructions: instructions,
            llm: model,
        });

        // STEP 7: Create the AgentSession.
        // Must pass at least {} — constructor destructures opts and crashes if undefined.
        const session = new voice.AgentSession({});

        // STEP 8: Cleanup handler
        let isCleaningUp = false;
        const handleCleanup = async () => {
            if (isCleaningUp) return;
            isCleaningUp = true;

            console.log("=".repeat(60));
            console.log("CLEANUP: Stopping recording and saving transcripts...");
            console.log("=".repeat(60));

            // Stop the LiveKit egress recording
            try {
                const egress = await recorder.stopRecording();
                console.log("✅ Recording stopped. Status:", egress?.status);
            } catch (e) {
                console.error("Error stopping recording:", e.message);
            }

            // Upload transcripts to S3 (in-memory, no local files)
            try {
                const urls = await transcription.uploadTranscriptsToS3(candidateId, uniqueSessionId);
                if (urls) {
                    console.log("✅ Transcripts uploaded to S3 successfully.");
                }
            } catch (e) {
                console.error("Error uploading transcripts:", e.message);
            }

            // Notify frontend that interview is done
            try {
                if (ctx.room.localParticipant) {
                    const payload = JSON.stringify({
                        type: "INTERVIEW_ENDED",
                        status: "completed",
                        room_id: ctx.room.name,
                        egress_id: recorder.egressId || uniqueSessionId,
                        message: "Interview completed. Recording is processing."
                    });
                    await ctx.room.localParticipant.publishData(
                        new TextEncoder().encode(payload),
                        { topic: "interview_status" }
                    );
                    console.log("✅ Sent INTERVIEW_ENDED to frontend.");
                }
            } catch (e) {
                console.error("Failed to send INTERVIEW_ENDED packet:", e.message);
            }
        };

        // STEP 9: Listen for user-triggered end-call signal
        ctx.room.on(RoomEvent.DataReceived, (payload, _participant, _kind, topic) => {
            if (topic === "user_actions") {
                try {
                    const data = JSON.parse(new TextDecoder().decode(payload));
                    if (data.type === "USER_ENDED_CALL") {
                        console.log("Received USER_ENDED_CALL — starting cleanup");
                        handleCleanup().then(() => ctx.room.disconnect()).catch(() => ctx.room.disconnect());
                    }
                } catch (e) {
                    console.error("Failed to parse data packet:", e.message);
                }
            }
        });

        // Fallback: if frontend disconnects first, run cleanup
        ctx.room.on(RoomEvent.Disconnected, () => {
            console.log("Room disconnected — running cleanup fallback");
            handleCleanup();
        });

        // STEP 10: Transcript capture — user speech
        session.on('user_input_transcribed', (ev) => {
            if (ev.transcript && ev.isFinal) {
                console.log(`USER: ${ev.transcript}`);
                transcription.onTranscript({ speaker: candidateId, text: ev.transcript, isFinal: true });

                if (ctx.room.localParticipant) {
                    ctx.room.localParticipant.publishData(
                        new TextEncoder().encode(JSON.stringify({
                            type: 'TRANSCRIPT', speaker: 'user',
                            text: ev.transcript, timestamp: new Date().toISOString()
                        })),
                        { reliable: true, topic: "transcripts" }
                    );
                }
            }
        });

        // STEP 11: Transcript capture — bot speech
        // ChatMessage has NO .text property; use .textContent getter
        session.on('conversation_item_added', (ev) => {
            const itemText = ev.item?.textContent;
            if (ev.item?.role === 'assistant' && itemText) {
                console.log(`BOT: ${itemText}`);
                transcription.onTranscript({ speaker: 'AI_Interviewer', text: itemText, isFinal: true });

                if (ctx.room.localParticipant) {
                    ctx.room.localParticipant.publishData(
                        new TextEncoder().encode(JSON.stringify({
                            type: 'TRANSCRIPT', speaker: 'bot',
                            text: itemText, timestamp: new Date().toISOString()
                        })),
                        { reliable: true, topic: "transcripts" }
                    );
                }
            }
        });

        // Log session errors
        session.on('error', (ev) => {
            console.error("❌ Session error:", ev);
        });

        // STEP 12: Start the agent session (connects Gemini Live to the room)
        console.log("Starting agent session...");
        try {
            await session.start({ agent, room: ctx.room });
            console.log("✅ Agent session started.");
        } catch (e) {
            console.error("❌ Failed to start agent session:", e);
            return;
        }

        // STEP 13: Start LiveKit cloud recording (Egress → S3)
        // Start recording before triggering the greeting so the opening
        // words are captured.
        try {
            await recorder.startRecording(ctx.room.name, candidateId, uniqueSessionId);
            console.log("✅ Cloud recording started. Egress ID:", recorder.egressId);
        } catch (e) {
            console.error("⚠️  Could not start cloud recording:", e.message);
            // Interview continues without recording
        }

        // STEP 14: Trigger the bot's opening greeting.
        // Gemini Live does NOT speak automatically on connect — generateReply()
        // kicks off the first turn so the bot greets the candidate.
        try {
            session.generateReply();
            console.log("✅ Initial greeting triggered.");
        } catch (e) {
            console.error("⚠️  Could not trigger initial greeting:", e.message);
        }

        // STEP 15: Auto-end interview after time limit
        const startTime = Date.now();
        const interval = setInterval(async () => {
            const elapsed = (Date.now() - startTime) / 1000;
            if (elapsed >= interviewDuration * 60) {
                console.log("⏰ Interview time limit reached");
                try {
                    session.say("Thank you for your time! The interview has concluded. Have a great day!");
                } catch (e) { /* ignore */ }
                setTimeout(async () => {
                    await handleCleanup();
                    await ctx.room.disconnect();
                    clearInterval(interval);
                }, 5000);
            }
        }, 30000);

        console.log("=".repeat(60));
        console.log("✅ INTERVIEW AGENT IS LIVE AND READY");
        console.log("=".repeat(60));
    },
});

// Run the agent worker only when this file is executed directly.
if (process.argv[1] === fileURLToPath(import.meta.url)) {
    cli.runApp(new WorkerOptions({
        agent: fileURLToPath(import.meta.url),
        // Small cloud instances (Render Starter = 0.5 CPU) are slow to spin up job
        // subprocesses — each imports aws-sdk + the Google plugin + the rtc-node
        // native addon. Production defaults (3 prewarmed procs, 10s init timeout)
        // cause "runner initialization timed out" there, so relax both.
        numIdleProcesses: 1,
        initializeProcessTimeout: 60_000,
        // In production the worker's health server binds a hardcoded 8081 and
        // ignores $PORT. On hosts that route to an injected port (Render/Railway
        // web services), bind that instead so health checks pass. Falls back to
        // the framework default when PORT is unset (e.g. a background worker).
        ...(process.env.PORT ? { port: Number(process.env.PORT) } : {}),
    }));
}
